### ISSUE TITLE
Legion: Extending layout constraints for polymorphic and variadic tasks

### ISSUE THREAD

**Issue by magnatelee** [StanfordLegion/legion#855]
Many places in the layout constraint code expect concrete field IDs, which make them difficult to use in cases where field IDs are generated dynamically. Furthermore, there are now many examples where tasks are field-polymorphic (in the sense that they are agnostic to the field IDs that are passed in), polymorphic, or variadic, and the current monomorphic layout constraints cannot be used to constraint such tasks. To this end, I'd like to propose the following changes to the current layout constraints:

- We replace every use of `FieldID` with `FieldOf`, a field descriptor object. A `FieldOf` can be one of a concrete field ID (`Concrete{fid}`), an index within a region requirement (`Position{i,j}`), and a predicate on the mapping tag (`Pred{tag,mode}`):

    * A descriptor `Position{i, j}` denotes the `j`-th instance field of the `i`-th requirement for a given task invocation. The indices `i` and `j` can be some predefined value `FORALL`, which selects the entire set. For example, `Position{0,FORALL}` denotes all instance fields of region requirement `0`. 

    * A descriptor `Pred{t,m}` denotes any region requirement whose tag value `tag` matches with `t`. The mode `m` determines how the match is checked; if `m` is `EQ`, the tag must exactly match (i.e., `tag == t`). If `m` is `MASK`, the match is determined by `tag & t > 0`. This is to exploit the domain knowledge that may already be available in the program; with variadic tasks, the program would already be passing the kind of data that each region requirement holds in the mapping tags.

    * When a descriptor that selects multiple fields in a constraint expecting a single field ID, it will lift the constraint to a list of constraints at substitution time. For example, `AlignmentConstraint{ Position{0,FORALL}, =, 64 }` will be flattened into a list of `AlignmentConstraint{ Concrete{fid1}, =, 64 }`, `AlignmentConstraint{ Concrete{fid2}, =, 64 }`, etc., where `fid`, `fid2`, etc. are the fields of the region requirement `0`.

- We add a "default" layout constraint set to [`TaskLayoutConstraintSet`](https://github.com/StanfordLegion/legion/blob/stable/runtime/legion/legion_constraint.h#L687-L697) and an optional parameter `conjoin` to decide how this default is used. If `conjoin` is true, the constraint applied to a region requirement `i` is a conjunction of the default and `layouts[i]`. Otherwise, the default will be overriden by `layouts[i]`. 

- We introduce a method `TaskLayoutConstraintSet::substitute` that converts a `TaskLayoutConstraintSet` with any field descriptors other than `Concrete` and/or a default layout into one that has none.

This is by no means a final design and I’d like your comments on this.

(There is a big question of how we make sure the constraints are internally consistent, but that is a problem already without this extension, so let’s have that discussion in another occasion.)

**Comment by lightsighter** [comment]
For everyone's visibility, in general I'm in favor of the proposal and what it will accomplish. Here are some specific comments and questions.

> A descriptor Position{i, j} denotes the j-th instance field of the i-th requirement for a given task invocation. 

How are you going to order the fields on a region requirement to index the `j`-th one? There are both the `privilege_fields` (set) and `inst_fields` (vector) data structures on region requirements, but only `privilege_fields` are required now. Would users have to use the `inst_fields` in order to use layout constraints?

> When a descriptor that selects multiple fields in a constraint expecting a single field ID, it will lift the constraint to a list of constraints at substitution time.

Isn't this an implementation detail and not something that is exposed to the user? I believe that we do all constraint satisfaction testing inside the runtime. When do you think we actually return a "concrete" set of layout constraints that the user will actually want to use?

> We add a "default" layout constraint

So far, the absence of layout constraints in the `TaskLayoutConstraintSet` data structure has been the "default". The runtime never tests any region requirements against constraints that aren't present. Instead the test goes the other direction: only constraints explicitly specified in the `TaskLayoutConstraint` data structure are used to evaluate whether the variant can be used.

> We introduce a method TaskLayoutConstraintSet::substitute 

This comes back to my earlier question as well: why would users ever need a "concrete" version of  a `TaskLayoutConstraintSet` object? The only place we use those is inside the runtime to make sure that variants are being used on instances with the correct layout in which case we're always testing a potentially polymorphic `TaskLayoutConstraintSet` data structure against a set of instances with concrete layout constraints. In general I think the rule should always be the polymorphic constraints can go into the runtime, but concrete layout constraints always come out of the runtime. So the application can pass in polymorphic constraints for `TaskLayoutConstraintSet` data structures when registering variants or for `LayoutConstraintSet` data structures when looking for instances. The only constraints the runtime ever returns are `LayoutConstraintSet` objects for concrete instances which are never polymorphic. Is there another use case that I'm missing?

**Comment by magnatelee** [comment]
> How are you going to order the fields on a region requirement to index the j-th one? There are both the privilege_fields (set) and inst_fields (vector) data structures on region requirements, but only privilege_fields are required now. Would users have to use the inst_fields in order to use layout constraints?

Yes, the `j` was meant to be an index to the `inst_fields`. I was expecting this to be okay, since `RegionRequirement::add_field` adds a field to both `privilege_fields` and `inst_fields`, unless the caller explicitly sets not to do so, and most cases I've seen fall to this default behavior. Unless we change `privilege_fields` to a vector, I can't think of any other way to select field IDs by positions.

An alternative approach I can think of is to use predicates that select field IDs based on their semantic info (as field IDs don't have mapping tags attached).

> Isn't this an implementation detail and not something that is exposed to the user? I believe that we do all constraint satisfaction testing inside the runtime. When do you think we actually return a "concrete" set of layout constraints that the user will actually want to use?

I think the user might want to query the `TaskLayoutConstraintSet` in the mapper to see what instances the mapper is supposed to create. (The easiest way to create an instance that satisfies the constraint is to pass exactly that constraint to the runtime when creating the instance.) So the user may need to know the semantics of `FORALL` in places for a single field ID.

> So far, the absence of layout constraints in the TaskLayoutConstraintSet data structure has been the "default". The runtime never tests any region requirements against constraints that aren't present. Instead the test goes the other direction: only constraints explicitly specified in the TaskLayoutConstraint data structure are used to evaluate whether the variant can be used.

I forgot to mention that this default constraint is optional, which will give you back the underspecification-by-default that you want. The default constraint is useful for variadic tasks, since it is a for-all constraint insensitive to the arity of the task. 

> This comes back to my earlier question as well: why would users ever need a "concrete" version of a TaskLayoutConstraintSet object? The only place we use those is inside the runtime to make sure that variants are being used on instances with the correct layout in which case we're always testing a potentially polymorphic TaskLayoutConstraintSet data structure against a set of instances with concrete layout constraints. In general I think the rule should always be the polymorphic constraints can go into the runtime, but concrete layout constraints always come out of the runtime. So the application can pass in polymorphic constraints for TaskLayoutConstraintSet data structures when registering variants or for LayoutConstraintSet data structures when looking for instances. The only constraints the runtime ever returns are LayoutConstraintSet objects for concrete instances which are never polymorphic. Is there another use case that I'm missing?

The `substitute` function isn't a user-facing function but a function used inside the runtime. I was just explaining what this function is supposed to do. But, like I said above, the user may want to retrieve a concrete `TaskLayoutConstraintSet` inside the mapper.

**Comment by lightsighter** [comment]
> Yes, the j was meant to be an index to the inst_fields.

I think this is fine, but it does mean that we need to be upfront with users that we are indexing into the `inst_fields` data structure here and it's up to them to make sure that it is populated since the rest of the runtime will not check that.

> I think the user might want to query the TaskLayoutConstraintSet in the mapper to see what instances the mapper is supposed to create. 

Sure, but that query call will be back into the runtime, so the user won't actually be doing the substitution themselves, but instead it will be handled inside the runtime as part of the constraint evaluation. I think my point is that they don't actually manipulate those `TaskLayoutConstraintSet` directly, but instead use it as an argument back into the runtime.

> I forgot to mention that this default constraint is optional, which will give you back the underspecification-by-default that you want.

Ok, that sounds fine. I would like to maintain this invariant that the default constraints are underspecified rather than having to specify some kind of default constraints that users need to know about.

> The substitute function isn't a user-facing function but a function used inside the runtime. I was just explaining what this function is supposed to do.

I think we are on the same page. I think it's just useful to distinguish what the user sees from what we are going to implement inside the runtime as it makes it more clear what the interface for the user is actually going to be.

**Comment by mpokorny** [comment]
A brief question, because this enhancement might prove rather useful to me, but it's possible that I don't understand it correctly. Is the predicate for tag matching actually applied to "any region requirement", or is there a dependence on a field space? The former would be far more useful to me than the latter, and it seems that you've described the former -- is that right?

**Comment by lightsighter** [comment]
I believe that @magnatelee was intending for the predicates to apply on a per-region-requirement basis, because they should be applying to `MappingTagID` for a the region requirement to which the layout constraints apply. Just for reference, currently you can describe one or more `LayoutConstraintSet` objects for each region requirement (applying to different fields of that region requirement):
https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/legion/legion_constraint.h#L743
So these predicates would be a part of the `LayoutConstraintSet` object (and it's sub-constraints) and therefore only be tested against a specific region requirement.
